src/cfg/File.h

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (C) 1996-2021 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#ifndef _SQUID__SRC_CFG_FILE_H
+#define _SQUID__SRC_CFG_FILE_H
+
+#include "base/Here.h"
+#include "cfg/forward.h"
+#include "sbuf/SBuf.h"
+
+#include <cstdio>
+#include <string>
+
+namespace Cfg
+{
+
+/**
+ * Class used to store required information for the current
+ * configuration file.
+ */
+class File
+{
+public:
+    File(const char *path) : filePath(path) {}
+    ~File();
+
+    /// \return true if the configuration file is open
+    bool isOpen() const { return bool(fd); }
+
+    /// \return the next line to be parsed from this file
+    SBuf nextLine();
+
+    /// \return the configuration file name and line number being processed
+    SourceLocation lineInfo() const { return SourceLocation("parsing", filePath.c_str(), lineNo); }
+
+    /// open and load contents from the file
+    void tryLoadFile();
```

Comment:
Why is there a public method that triggers what any usable File should do -- loading the file? Should not this method be called from the File _constructor_?

And please drop the "try" prefix. All methods may throw by default. There is nothing special about this one AFAICT, and this method does not have an exception processing wrapper method that sometimes justifies such "try" prefixes.

Reply:
This is separate from the constructor because there are currently many configuration parsers in Squid code.
 * at least one of them initiates file loading from global scope separately from loading the filename and construction a Cfg::File object.
 * at least one of them constructs a Cfg::File using a filename of a pipe or socket.

The "try" was there because the method may not load anything "successfully" (data.st_size == 0) and also may be repeated if the callers or configurator has knowledge the Cfg::File object is not party to (eg this is a socket/pipe that just got new data, file that got edited).  I have renamed to load() in 616877b.